### PR TITLE
MPR#7528: Format.pp_set_geometry

### PR DESCRIPTION
**Pull request by Octachron**
[MPR7528](https://caml.inria.fr/mantis/view.php?id=7528):
This PR proposes to add a new `pp_set_geometry: formatter -> max_indent:int -> margin:int -> unit ` function to the Format module and its stdout `set_geometry` and `get` counterparts.

This new function sets both the margin and the maximum indentation limit. The labeled argument are here to avoid having two indistinguishable integer arguments.

If `0 < max_indent < margin`, `pp_set_geometry ppf ~max_indent ~margin` is equivalent to
```Ocaml
pp_set_margin ppf margin; pp_set_max_indent ppf max_indent
```
After such call, the margin and maximum indentation limit are guaranteed to be set to the requested values. 
Outside of this cone `0 < max_indent < margin`, the function raises an invalid argument exception.

The main motivation behind this new function is to avoid the memory effects that plague `set_margin` and `set_max_indent` when used separately.

In particular, even if `0 < max_indent < margin`, reversing the order of the call to `set_max_indent`
and `set_margin`, i.e.
```OCaml
set_max_indent max_indent; set_margin margin
```
may not set the maximum indentation to `max_indent`:
if the current margin `get_margin ()` is less than `max_indent`, the call to `set_max_indent` fails silently and let unchanged the current value of the maximum indentation limit.
Even worse, if `max_indent > margin`, the resulting formatter state may invisibly depend on the order of the calls to `set_margin` and `set_max_indent`, the current margin, and the current maximum indentation limit. Similarly, setting only the margin with `set_margin` can modify the maximum indentation limit and the resulting maximum indentation limit depends on the history of the calls to `set_margin`, e.g. `set_margin 120; set_margin 60; set_margin 50` leads to a different state than `set_margin 90; set_margin 60; set_margin 50`.

I made the choice of raising an exception due to the difficulty of defining a completely non-surprising projector on the cone `0< max_indent < margin` . In particular, for `max_indent >= margin > 1`, what is the nearest point on the cone? Is it `(margin-1,margin)`, or `(max_indent, max_indent + 1)`, or maybe the middle point `( (max_indent + margin - 1) / 2, (margin + max_indent + 1) /2 )`?

For reference, in #1236 @jberdine alternatively suggested to have a `pp_set_geometry_with_offset` function (with a better name) that takes the  offset `margin - max_indent` and `margin` as argument; but I thought it will be clearer with the existing documentation to keep `max_indent` as an argument.

### CODE REVIEW

**Comment by gasche** [comment]
cc @pierreweis, @rbonichon: any feedback is welcome

**Comment by Drup** [comment]
I think this is a good idea. We might even want to (*gasp*) deprecate the other ones.

**Comment by rbonichon** [comment]
This is a joint response with @pierreweis.

## About the proposal itself

The idea to set margin and max_indent simultaneously is not a bad one. 
We would favor a record like so:

>    type geometry = {
       max_index: int;
       margin: int;
     }

This would ensure distinguishability not only for the input parameters (what you used labels for) but also for the results of get_geometry.

What we want is something that **never** fails: a satisfying solution would be either do what was specified, if possible, or just do nothing.

Thus the addition of the functionality would amount to just sequencing the two operations, which ensures that margin is always set before max_indent, whatever the values of max_indent and margin. 


## About failing silently

There are moments where you cannot put in place the geometry required by the
user. In this case, it is not an option to fail. 

It was indeed on purpose: `Format` never prints error messages nor raises
exceptions. Indeed, raising exceptions would break internal invariants.
And printing warnings/error messages would badly interfere with user
outputs. 

Actually it does not fail silently: it keeps the invariants of the formatter intact.

## Memory effects

You indeed have brutal (memory) side effects when using `pp_set_margin` or
`pp_set_max_indent`. Actually whenever `margin - indent > 1`, `pp_rinit` is
 called and the whole state is reset.
   
Whenever you set new margin or maximum indentation, you just break all the
hypotheses that allowed to format the past output in the first hand. The
formatter is thus forced to invalidate whatever is in the queue that was not
already displayed. You could even lose some material in the process.

(Maybe a flush would be an enhancement there).

The sequence of  different set_margin invocations of course does not lead to the same result.
You are just mutating the internal state of the formatter. What else would you expect ?
 Format is indeed an extremely stateful module: it has pros and cons.

## Bugs

We have found some discrepancies with the doc thanks to your PR. The doc
states that nothing happens if [d] is smaller than 2 for
pp_set_max_indent. The code does not ensure that.

**Comment by Drup** [comment]
> We would favor a record like so [ .... ]

While returning a record is fine, taking one as argument is less so.
The advantage of the labels is that they are relatively easy to extend in a backward compatible manner. This is not the case with records. You already broke compat for this exact reason once, I think we should do the right thing directly this time around. :)


> What we want is something that never fails: a satisfying solution would be either do what was specified, if possible, or just do nothing.

I'm sorry, but that's just a terrible design pattern. If you need to learn something from the last 20 years of API design, is that failure should never, ever, *ever* be silent. Raising `invalid_argument` here does not break any internal invariant of Format, and even if it did, it doesn't matter: Programmers are not supposed to catch this exception and keep going. They are supposed to fix their program.

I agree that flushing on margin change is probably a good idea.

**Comment by Octachron** [comment]
First, thanks for the review.

>We would favor a record like so:
>
>   type geometry = {
>    max_index: int;
>    margin: int;
>    }

Returning a geometry record for `get_geometry` seems like a clear improvement. I am not sure about having to construct a record for using `set_geometry`.

>  What we want is something that never fails: a satisfying solution would be either do what was > specified, if possible, or just do nothing.

Could you please elaborate on why you are so intent on never raising an exception? If the requested geometry is invalid, not raising an exception means that the user expectation that the geometry had been set to the required value is broken. I choose to raise an exception to make sure that the library user can fix its broken code as early as possible and can rely on the sensible expectation that after a `set_geometry geometry` the actual value of the geometry is guaranteed to be `geometry`.

 If the function were to do nothing in presence of invalid geometry, I believe that are the very least it should be modified to return a `(unit,unit) result` error code. However, error code are far too easy to ignore from my point of view.

> it keeps the invariants of the formatter intact.

Keeping the internal invariant of the formatter but breaking user invariant is failing as far as I am concerned.

> The sequence of different set_margin invocations of course does not lead to the same result.
You are just mutating the internal state of the formatter. What else would you expect ?
Format is indeed an extremely stateful module: it has pros and cons.

I would expect that when exposing a given coordinate chart to an user for a parameter space, changing individual coordinate never ever changes one of the other coordinates. Of course, this is not possible with the coordinate chart exposed by the Format module, since the parameter space is a manifold with boundary, and the exposed coordinate chart is not adapted to the boundary. In spite of this difficulty, at the very least, I would expect that the position on the parameter state can be deduced from the requested coordinate (i.e. `set_margin y`) and the current value of the remaining coordinates (here the maximum indentation limit). However, `Format`fails to satisfy even this weak property since the results depends on the previous value of the coordinate being set.

**Comment by rbonichon** [comment]
> that failure should never, ever, ever be silent.

If the specification explicitly states that it does nothing, why would you consider a failure the fact of implementing the spec ?

> Raising invalid_argument here does not break any internal invariant of Format, and even if it did, it doesn't matter: 

Yes it breaks *all* the invariants since it will let the formatter in an unknown state.
We beg to disagree and think it matters that this above should be prevented

**Comment by rbonichon** [comment]
> Could you please elaborate on why you are so intent on never raising an exception?

Once upon a time, there were exceptions all over the Format module. An exception was raised when trying to close non-opened boxes, another one when printing when no box was opened yet, etc. This was causing a lot of problems, the worst being that the OCaml compiler itself was crashing when trying to print error messages.
Most users were thus extremely annoyed at this behavior. So, exceptions were removed from Format and the current defensive behavior was implemented. Users could not crash the formatter, even if they were misusing it. 

> Keeping the internal invariant of the formatter but breaking user invariant is failing as far as I am concerned.

I am not sure what you are referring to when you speak of user invariant. I guess you mean "user expectations". In this case, this can be fixed by reading the documentation.

> I would expect that when exposing a given coordinate chart to an user for a parameter space, changing individual coordinate never ever changes one of the other coordinates

In this case margin and max_indent are not "free parameters" but rather linked together x and y point coordinates on a circle. The relations are enforced inside the module as invariants, and the documentation should reflect that as well.

One solution of our disagreement can be to provide a predicate that would tell the user if its proposed geometry is valid. What do you think ?

**Comment by Octachron** [comment]
> Once upon a time, there were exceptions all over the Format module. An exception was raised when trying to close non-opened boxes, another one when printing when no box was opened yet, etc.

All these exceptions depend on the previous state of the format and not only on the argument provided to the current function call. This is not the case with `pp_set_geometry` where knowing the value of the arguments is enough to determine if an exception would be raised or not. I believe this difference does matter in the context of raising an exception or not.

>I am not sure what you are referring to when you speak of user invariant. I guess you mean "user expectations". In this case, this can be fixed by reading the documentation.

Let's say user expectations. You cannot fix user expectations with documentation (except, maybe, with a stellar documentation that redefine the underlying concepts in the user mind), and the expectation that a setter function does set is quite widespread.

> In this case margin and max_indent are not "free parameters" but rather linked together x and y point coordinates on a circle.

Indeed the parameter space is not a vector space but a manifold with boundary. However, for a cone, there are few global coordinate charts were the boundary can be defined by ` x = 0 ||  x = x_max | y = 0  ` and thus modifying one coordinate do not have to change any of the other coordinates because coordinate curves simply never cross the boundary of the manifold . Anyway, it is too late to change the coordinate chart exposed by Format.

> One solution of our disagreement can be to provide a predicate that would tell the user if its proposed geometry is valid. 

I have nothing against a geometry validation function, which could be useful for other libraries; but I have a counter-proposition: a pair of function `pp_set_geometry`and `pp_safe_set_geometry` (or `pp_set_geometry_if_valid`) .

**Comment by rbonichon** [comment]
What would the specification `pp_safe_set_geometry` be ? If this is: "set geometry only if it is valid, otherwise do nothing", I think this might end the discussion with a satisfactory agreement.

Regarding exceptions in `Format`, there are, as I see it, 3 choices:
1. Never raise exceptions and always try to do something meaningful in the face of weird user specification -- I would call it the current choice
2. Always raise exceptions on user errors - been there done that : it makes the module unusable for most people
3. Sometimes raise exceptions, sometimes not: I would argue that it lacks coherence (not that it never happens in the ocaml compiler) and it also paves the way to choice 2. In all honesty, why not add other exceptions if one already is raised somewhere ? `set_margin` and `set_max_indent` may seem somewhat different than, say, the printing stack, but actually, they are as much part of the state --- as you mentioned, sequences of `set_margin` with different arguments may produce different `max_indent` in the end.

As an aside, concerning raising exceptions, they may be somewhat problematic in a pretty-printing context. Say you raise an exception for misuse: what happens if you catch it and you try to print (with `Format` of course) a nice error message ? What of the user when you raise an exception while he was trying to print a debug message ? 

> You cannot fix user expectations with documentation (except, maybe, with a stellar documentation that redefine the underlying concepts in the user mind), and the expectation that a setter function does set is quite widespread.

True but it also widespread to expect the user to read the documentation before using an API, especially for a module like `Format` which uses a lot of specific concepts. Maybe the name `set_*` is not the best for these functions. In a way, it behaves more like `mk_margin` and `mk_max_ident`...

**Comment by gasche** [comment]
Four unordered points:

- Would it be ok to raise an exception if the print queue on the formatter that received the invalid instruction was systematically flushed before that?

- I don't expect `pp_set_{margin,max_indent,geometry}` to be called in the middle of printing operations (does it even have a well-defined semantics in this case?). Maybe it would make sense to have a different error-handling regime for such "toplevel" functions (also: `set_formatter_out_channel`, etc.) than for the functions used for printing in the stack.

- If there is a coherent design choice in the current API to avoid raising exceptions, I think it makes sense to understand why and stick to that choice as long as a coherent counter-proposal that solves the actual problem (and that seems fairly hard to solve) is made -- maybe as a different library.

- This PR may not be the best place to start an overhaul of the APi. If I was the submitter, I would rather align myself to the recommendations of the maintainer on this matter and try get the PR merged -- the discussion can continue in this issue after merging if desired.

**Comment by Octachron** [comment]
> What would the specification pp_safe_set_geometry be ? If this is: "set geometry only if it is valid, otherwise do nothing", I think this might end the discussion with a satisfactory agreement.

Exactly, let's settle for this.

> As an aside, concerning raising exceptions, they may be somewhat problematic in a pretty-printing context. Say you raise an exception for misuse: what happens if you catch it and you try to print (with Format of course) a nice error message ? 

Speaking of specification, the specification of the `Invalid_argument` exception is that it indicates a programming error that should be fixed and this exception should not be caught (https://caml.inria.fr/pub/docs/manual-ocaml/core.html#sec549).

**Comment by Octachron** [comment]
As discussed, I have added a `safe_set_geometry` function that does nothing in presence of invalid geometry, a geometry argument returned by `get_geometry` and a validation function for geometries.

**Comment by gasche** [comment]
I'm afraid that this PR may have been forgotten for no good reason. Is there anything blocking a merge now?

**Comment by Octachron** [comment]
I would be nice to have this PR merged. People really tend to forget to set `max_indent` after setting the `margin`, and not only beginners: there was a hidden bug in the compiler itself due to this very issue that was temporarily exposed in #1952.

**Review by gasche — APPROVED** [review]
I'll take the responsibility to consider that the discussion has converged to a reasonable degree, and approve the PR.

**Comment by rbonichon** [comment]
@gasche @Octachron I agree that the discussion has converged adequately so that this PR can be merged.

**Pull request merged**
